docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

```diff
@@ -0,0 +1,1946 @@
+
+.. note::
+  If you still need to perform a query with an explicit package-id, you can either use
+  a previous version of PQS, or add the following filter predicate to your query:
+  ``SELECT \* FROM active('my_package:My.App:MyTemplate') WHERE package_id = 'my_package_id'``
+
+Given that PQS uses a document-oriented model for ledger content
+(JSONB), extensions to contract payloads are handled simply by returning
+the additional data in the blob.
+
+Daml Shell
+----------
+
+Daml Shell builds on PQS by providing a shell interface to inspect the
+ledger using package-name to view all versions of contracts, in an
+integrated way.
+
+Daml-Script-Beta
+----------------
+
+Daml 2.9 introduces a new version of Daml Script, which can be used by
+depending on ``daml-script-beta`` in your ``daml.yaml``, as you will have seen
+in `Writing your first upgrade <#writing-your-first-upgrade>`__. Only this version of Daml Script
+supports upgrades over the Ledger API.
+
+All commands and queries in this version of daml script will now use
+upgrades/downgrades automatically, to ensure that the correct versions
+of choices are exercises, and correct payloads are returned.
+
+The following additional functionality is available for more advanced
+uses of SCU.
+
+**Exact commands**
+
+Each of the 4 submission commands now have an “exact” variant, of the
+forms ``createExactCmd``, ``exerciseExactCmd``, ``exerciseByKeyExactCmd`` and
+``createAndExerciseExactCmd``. These commands will force the participant to
+use the exact version of the package that your script uses, this is most
+useful when you want to be absolutely certain of the choice code you are
+calling. Note that exact and non-exact commands can be mixed in the same
+submission.
+
+Daml Studio support
+-------------------
+
+Daml Studio runs a reference model of Canton called the IDE Ledger, this
+ledger has been updated to support most of the functionality of upgrades
+in daml-script. The behavior that this ledger does not implement is the
+following:
+
+-  Upgrades type checking
+   The Upgrades type errors you’ll get when running daml build are not
+   currently shown as code intelligence in Daml Studio.
+
+-  Contract ID verification
+   This may allow some contract metadata changes to take effect that
+   Canton otherwise would not.
+
+-  Per-submission package preference
+   Any submission to the IDE Ledger will mimic a Canton participants
+   default package preference of the most recent package version.
+
+Testing
+=======
+
+Upgrade validity checking
+-------------------------
+
+While the compile time upgrade checks are quite thorough, they are not
+yet a perfect mirror of the checks that a participant does when a
+package is uploaded. We recommend that as a final check for the validity
+of your upgraded package, you can either:
+
+-  Run a Canton sandbox (running ``daml sandbox``) and upload your old and
+   new package (``daml ledger upload-dar``).
```

Comment:
```suggestion
   new packages (``daml ledger upload-dar``).
```
There can be multiple packages. See an upgrade of CN here: https://github.com/DACH-NY/canton-network-node/commit/917a631b19a264f3a98957d9c9018e3ee9ae8d62#diff-d2555057df0d1d7e4b946dee2ceeabeac77c2d51c94f0912e02c6b119bd7ad12